pkg/agent/proxy/proxier.go

```diff
@@ -452,19 +452,29 @@ func (p *proxier) installServices() {
 				continue
 			}
 
-			// Install another group when Service externalTrafficPolicy is Local.
-			if p.proxyAll && svcInfo.NodeLocalExternal() {
+			if p.proxyAll {
 				groupIDLocal, _ := p.groupCounter.Get(svcPortName, true)
-				var localEndpointList []k8sproxy.Endpoint
-				for _, ed := range endpointUpdateList {
-					if !ed.GetIsLocal() {
+				// Install another group when Service externalTrafficPolicy is Local.
+				if svcInfo.NodeLocalExternal() {
+					var localEndpointList []k8sproxy.Endpoint
+					for _, ed := range endpointUpdateList {
+						if !ed.GetIsLocal() {
+							continue
+						}
+						localEndpointList = append(localEndpointList, ed)
+					}
+					if err = p.ofClient.InstallServiceGroup(groupIDLocal, svcInfo.StickyMaxAgeSeconds() != 0, localEndpointList); err != nil {
+						klog.ErrorS(err, "Error when installing Group for Service whose externalTrafficPolicy is Local")
 						continue
 					}
-					localEndpointList = append(localEndpointList, ed)
-				}
-				if err = p.ofClient.InstallServiceGroup(groupIDLocal, svcInfo.StickyMaxAgeSeconds() != 0, localEndpointList); err != nil {
-					klog.ErrorS(err, "Error when installing Group for Service whose externalTrafficPolicy is Local")
-					continue
+					// Uninstall the group with only local Endpoints when Service externalTrafficPolicy is changed from Local
+					// to Cluster.
+				} else if !svcInfo.NodeLocalInternal() && pSvcInfo != nil && pSvcInfo.NodeLocalExternal() {
+					if err := p.ofClient.UninstallServiceGroup(groupIDLocal); err != nil {
+						klog.ErrorS(err, "Failed to remove Group for Service with only local Endpoints", "Service", svcPortName)
+						continue
+					}
+					p.groupCounter.Recycle(svcPortName, true)
```

Comment:
The group ID would be leaked if the service's external policy is always cluster. It should either call `p.groupCounter.Get(svcPortName, true)` inside the "else" section or unconditionally call UninstallServiceGroup and Recycle as long as NodeLocalExternal is false. I feel the latter is more robust as it could work for the case that the previous round of handling fails.

Reply:
I got what you means. The latter way is more robust, but I think this will get lots of error log when Services with NodeLocalExternal false are deleted.

Reply:
I think the methods InstallServiceGroup, UninstallServiceGroup are idempotent. Why would there be error logs?

Reply:
Even if `UninstallServiceGroup` is not idempotent, I don't get why deleting Services with false NodeLocalExternal will get error log, could you explain?

Reply:
I thought you meant when deleting Services with false NodeLocalExternal, unconditionally call UninstallServiceGroup to delete the group with only local Endpoints even it doesn't exist. I thought of delete a not existing group will return an error.